src/aggregated_range_proof/dealer.rs

```diff
@@ -0,0 +1,174 @@
+use proof_transcript::ProofTranscript;
+use std::clone::Clone;
+use util;
+
+use super::messages::*;
+
+/// Dealer is an entry-point API for setting up a dealer
+pub struct Dealer {}
+
+impl Dealer {
+    /// Creates a new dealer with the given parties and a number of bits
+    pub fn new(
+        transcript: &mut ProofTranscript,
+        n: usize,
+        m: usize,
+    ) -> DealerAwaitingValues {
+        transcript.commit_u64(n as u64);
+        transcript.commit_u64(m as u64);
+        DealerAwaitingValues { n, m }
+    }
+}
+
+/// When the dealer is initialized, it only knows the size of the set.
+pub struct DealerAwaitingValues {
+    n: usize,
+    m: usize,
+}
+
+impl DealerAwaitingValues {
+    /// Combines commitments and computes challenge variables.
+    pub fn receive_value_commitments(
+        self,
+        transcript: &mut ProofTranscript,
+        vc: &Vec<ValueCommitment>,
+    ) -> (DealerAwaitingPoly, ValueChallenge) {
+        assert!(vc.len() == self.m);
+        let mut A = RistrettoPoint::identity();
+        let mut S = RistrettoPoint::identity();
+
+        for commitment in vc.iter() {
+            // Commit each V individually
+            transcript.commit(commitment.V.compress().as_bytes());
+
+            // Commit sums of As and Ss.
+            A += commitment.A;
+            S += commitment.S;
+        }
+
+        transcript.commit(A.compress().as_bytes());
+        transcript.commit(S.compress().as_bytes());
+
+        let y = transcript.challenge_scalar();
+        let z = transcript.challenge_scalar();
+
+        (DealerAwaitingPoly { n: self.n }, ValueChallenge { y, z })
+    }
+}
+
+pub struct DealerAwaitingPoly {
+    n: usize,
+}
+
+impl DealerAwaitingPoly {
+    pub fn receive_poly_commitments(
+        self,
+        transcript: &mut ProofTranscript,
+        poly_commitments: &Vec<PolyCommitment>,
+    ) -> (DealerAwaitingShares, PolyChallenge) {
+        // Commit sums of T1s and T2s.
+        let mut T1 = RistrettoPoint::identity();
+        let mut T2 = RistrettoPoint::identity();
+        for commitment in poly_commitments.iter() {
+            T1 += commitment.T_1;
+            T2 += commitment.T_2;
+        }
+        transcript.commit(T1.compress().as_bytes());
+        transcript.commit(T2.compress().as_bytes());
+
+        let x = transcript.challenge_scalar();
+
+        (DealerAwaitingShares { n: self.n }, PolyChallenge { x })
+    }
+}
+
+pub struct DealerAwaitingShares {
+    n: usize,
+}
+
+impl DealerAwaitingShares {
+    pub fn receive_shares(
+        self,
+        transcript: &mut ProofTranscript,
+        proof_shares: &Vec<ProofShare>,
+        gen: &GeneratorsView,
+        y: Scalar,
+    ) -> Proof {
+        let value_commitments = proof_shares
+            .iter()
+            .map(|ps| ps.value_commitment.V.clone())
+            .collect();
+        let A = proof_shares
+            .iter()
+            .fold(RistrettoPoint::identity(), |A, ps| {
```

Comment:
this seems like a reason to implement `Sum` and `Product` upstream in `curve25519-dalek`